### PR TITLE
OCPQE-15045: Add dual stack vmedia IPI on BM

### PR DESCRIPTION
**Pull request by jadhaj**
The PR aims to add a test & job config for dual stack OpenShift deployment. For this purpose, a new workflow is introduced and a new step for configuring dual vips on install-config.yaml.

The job config utilize changes made recently to reserve-hosts.sh to dynamically retrieve IPv6 vips. necessary changes were also made to load balancer and DNS steps to adapt for dual stack.

### CODE REVIEW

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by sgoveas** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.12-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
@aleskandro Afer adding below configurations to WF, I can ping & ssh to nodes using IPv6:
1. echo -e "dhcp-option=option6:router,[fd99:2222:3456::1]" >> /opt/dnsmasq/hosts/optsdir/"${3}"
2.  auto-gateway: true
     auto-routes: true
     autoconf: true
     auto-dns: true

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

File: ci-operator/step-registry/baremetal/lab/ipi/install/baremetal-lab-ipi-install-commands.sh [hunk 1]
```diff
@@ -100,8 +100,10 @@ platform:
   baremetal:
     libvirtURI: >-
       qemu+ssh://root@${AUX_HOST}:$(sed 's/^[%]\?\([0-9]*\)[%]\?$/\1/' < "${CLUSTER_PROFILE_DIR}/provisioning-host-ssh-port-${architecture}")/system?keyfile=${CLUSTER_PROFILE_DIR}/ssh-key&no_verify=1&no_tty=1
-    apiVIP: $(yq ".api_vip" "${SHARED_DIR}/vips.yaml")
-    ingressVIP: $(yq ".ingress_vip" "${SHARED_DIR}/vips.yaml")
+    apiVIPs:
```
Comment (aleskandro) [inline ci-operator/step-registry/baremetal/lab/ipi/install/baremetal-lab-ipi-install-commands.sh:103]: Why do we need this if there is a new step that creates a patch for the install-config (used a few lines below)?

Comment (jadhaj) [reply ci-operator/step-registry/baremetal/lab/ipi/install/baremetal-lab-ipi-install-commands.sh]: I will revert and test, but I remember when running w/o touching this part, vips did not look good on install-config

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

File: ci-operator/step-registry/baremetal/lab/pre/dhcp-conf/baremetal-lab-pre-dhcp-conf-commands.sh [hunk 2]
```diff
@@ -62,4 +62,5 @@ timeout -s 9 10m ssh "${SSHOPTS[@]}" "root@${AUX_HOST}" bash -s -- \
   "'${DHCP_CONF}'" "'${DHCP_CONF_OPTS}'" "'${CLUSTER_NAME}'" <<'EOF'
   echo -e "${1}" > /opt/dnsmasq/hosts/hostsdir/"${3}"
   echo -e "${2}" > /opt/dnsmasq/hosts/optsdir/"${3}"
+  echo -e "dhcp-option=option6:router,[fd99:2222:3456::1]" >> /opt/dnsmasq/hosts/optsdir/"${3}"
```
Comment (aleskandro) [inline ci-operator/step-registry/baremetal/lab/pre/dhcp-conf/baremetal-lab-pre-dhcp-conf-commands.sh:65]: This can't stay here.

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

File: ci-operator/step-registry/baremetal/lab/pre/dhcp-conf/baremetal-lab-pre-dhcp-conf-commands.sh (hunk 2 above)
Comment (jadhaj) [reply ci-operator/step-registry/baremetal/lab/pre/dhcp-conf/baremetal-lab-pre-dhcp-conf-commands.sh]: Removed

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

**Comment by jadhaj** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-baremetal-ipi-ovn-dualstack-amd-vmedia-f7

File: ci-operator/step-registry/baremetal/lab/ipi/conf/load-balancer/baremetal-lab-ipi-conf-load-balancer-commands.sh [hunk 3]
```diff
@@ -47,21 +49,24 @@ listen api-server-6443
...
     mode tcp
     balance source
     server api_vip $INGRESS_VIP:80 check inter 1s
+    server api_vip_v6 [$INGRESS_VIP_V6]:80 check inter 1s
 listen ingress-router-443
     bind :::443
     mode tcp
     balance source
     server api_vip $INGRESS_VIP:443 check inter 1s
+    server api_vip_v6 [$INGRESS_VIP_V6]:443 check inter 1s
```
Comment (sgoveas) [inline ci-operator/step-registry/baremetal/lab/ipi/conf/load-balancer/baremetal-lab-ipi-conf-load-balancer-commands.sh:69]: We should check if ipv4 or v6 is enabled to add all these entries for the load balancer

File: ci-operator/step-registry/baremetal/lab/pre/dns/baremetal-lab-pre-dns-commands.sh [hunk 4]
```diff
@@ -18,7 +18,7 @@ BASE_DOMAIN=$(<"${CLUSTER_PROFILE_DIR}/base_domain")
 # shellcheck disable=SC1090
 . <(yq e 'to_entries | .[] | (.key + "=\"" + .value + "\"")' < "${SHARED_DIR}"/external_vips.yaml)
 # shellcheck disable=SC2154
-if [ ${#api_vip} -eq 0 ] || [ ${#ingress_vip} -eq 0 ]; then
+if [ ${#api_vip} -eq 0 ] || [ ${#ingress_vip} -eq 0 ] || [ ${#api_vip_v6} -eq 0 ] || [ ${#ingress_vip_v6} -eq 0 ]; then
```
Comment (sgoveas) [inline ci-operator/step-registry/baremetal/lab/pre/dns/baremetal-lab-pre-dns-commands.sh:21]: Same as above, check if ipv4 or ipv6 is enabled before this check

**Pull request closed without merging**